lgc/elfLinker/ColorExportShader.h

```diff
@@ -0,0 +1,86 @@
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  ColorExportShader.h
+ * @brief LGC header file: The class to generate the color export shader used when linking a pipeline.
+ ***********************************************************************************************************************
+ */
+#pragma once
+
+#include "GlueShader.h"
+#include "lgc/state/PalMetadata.h"
+#include "lgc/state/PipelineState.h"
+
+namespace lgc {
+
+class LgcContext;
+
+// =====================================================================================================================
+// A color export shader
+class ColorExportShader : public GlueShader {
+public:
+  ColorExportShader(PipelineState *pipelineState, llvm::ArrayRef<ColorExportInfo> exports);
+  ~ColorExportShader() override {}
+
+  // Get the string for this glue shader. This is some encoding or hash of the inputs to the create*Shader function
+  // that the front-end client can use as a cache key to avoid compiling the same glue shader more than once.
+  llvm::StringRef getString() override;
+
+  // Get the symbol name of the main shader that this glue shader is prolog or epilog for.
+  llvm::StringRef getMainShaderName() override;
+
+  // Get the symbol name of the glue shader.
+  llvm::StringRef getGlueShaderName() override { return "color_export_shader"; }
+
+  // Get whether this glue shader is a prolog (rather than epilog) for its main shader.
+  bool isProlog() override { return false; }
+
+  // Get the name of this glue shader.
+  llvm::StringRef getName() const override { return "color export shader"; }
+
+  // Update the entries in the PAL metadata that require both the pipeline state
+  // and export info.
+  void updatePalMetadata(PalMetadata &palMetadata) override;
+
+protected:
+  // Generate the glue shader to IR module
+  llvm::Module *generate() override;
+
+private:
+  llvm::Function *createColorExportFunc();
+
+  // The information stored here is all that is needed to generate the color export shader. We deliberately do not
+  // have access to PipelineState, so we can hash the information here and let the front-end use it as the
+  // key for a cache of glue shaders.
+  llvm::SmallVector<ColorExportInfo, 8> m_exports;
+  ExportFormat m_exportFormat[MaxColorTargets]; // The export format for each hw color target.
+  // The encoded or hashed (in some way) single string version of the above.
+  std::string m_shaderString;
+  PipelineState *m_pipelineState; // The pipeline state.  Used to set meta data information.
+  bool m_killEnabled;             // True if this fragment shader has kill enabled.
```

Comment:
Can we initialize these members here and drop the `memset` from the constructor?

Reply:
Aren't are all of the already default initialized?  I think we can just drop the memset.  However, this PR is mean to just move code around.  I did change the return type of one function to return a unique pointer, but how much can I change and still claim this has no functional change?

I think I will leave this as it for this PR.

Reply:
> Aren't are all of the already default initialized? 
Not unless you have ` = ...` here or explicit initialization in the constructor initialization list. What I'm advocating for it setting all primitives and PODs at the definition (e.g., `int sth = 0;`) so that the code is obviously free from uninitialized variables without having to go over all constructors to confirm that.

+1. I saw you changed the return type and was not sure if you wanted clean up the code at the same time or not.

Reply:
Right, it will zero-initialize in the default constructor, but not in a user defined constructor.